### PR TITLE
Rewritten modal without jquery

### PR DESCRIPTION
**Pull request by alekitto**
To make it work correctly the following changes were required:

- Trigger events in jQuery if available
- Add util functions for manipulating data attributes and jQuery data object
- `Util.extend` now supports multiple objects

### CODE REVIEW

File: js/src/dom/eventHandler.js [hunk 1]
```diff
@@ -197,21 +198,18 @@ const EventHandler = {
...
-        for (const keyHandlers in storeElementEvent) {
-          if (!Object.prototype.hasOwnProperty.call(storeElementEvent, keyHandlers)) {
-            continue
-          }
-          // delete all the namespaced listeners
-          if (inNamespace && keyHandlers.indexOf(originalTypeEvent) > -1) {
-            const handlerFn = events[elementEvent][keyHandlers]
-            EventHandler.off(element, elementEvent, handlerFn.originalHandler)
-          }
+        const handlerKey = keyHandlers.replace(stripUidRegex, '')
```
Comment (Johann-S) [inline js/src/dom/eventHandler.js]: Why did you add a new regex here ?  `off` method already work

File: js/src/dom/eventHandler.js [hunk 2]
```diff
@@ -234,14 +232,31 @@ const EventHandler = {
...
     const isNative    = nativeEvents.indexOf(typeEvent) > -1
+
+    const $ = Util.jQuery
+    let jQueryEvent
+
+    let bubbles = true
+    let nativeDispatch = true
+    let defaultPrevented = false
+
+    if (!isNative && typeof jQuery !== 'undefined') {
```
Comment (Johann-S) [inline js/src/dom/eventHandler.js]: here you check for `jQuery !== undefined` it should be better with `$` which come from your `Util.jQuery` getter

File: js/src/dom/eventHandler.js [hunk 3]
```diff
@@ -250,7 +265,21 @@ const EventHandler = {
...
-    element.dispatchEvent(evt)
+
+    if (defaultPrevented) {
+      evt.preventDefault()
+    }
+
+    if (nativeDispatch) {
+      element.dispatchEvent(evt)
+    } else {
+      evt.stopImmediatePropagation()
```
Comment (Johann-S) [inline js/src/dom/eventHandler.js]: why do you call `stopImmediatePropagation` here ? `evt` isn't dispatched

File: js/src/util.js [hunk 4]
```diff
@@ -119,18 +136,54 @@ const Util = (() => {
...
+    setDataAttribute(element, key, value) {
+      const $ = this.jQuery
+      if (typeof $ !== 'undefined') {
+        $(element).data(key, value)
+      }
+
+      element.setAttribute(`data-${key.replace(/[A-Z]/g, (chr) => `-${chr.toLowerCase()}`)}`, value)
+    },
+
+    removeDataAttribute(element, key) {
```
Comment (Johann-S) [inline js/src/util.js]: `setDataAttribute` and `removeDataAttribute` should be in our DOM Manipulator here : https://github.com/twbs/bootstrap/blob/v4-without-jquery/js/src/dom/manipulator.js

File: js/src/modal.js [hunk 5]
```diff
@@ -537,53 +538,57 @@ const Modal = (() => {
...
-        if ($(this).is(':visible')) {
+      EventHandler.one(target, Event.HIDDEN, () => {
+        if (Util.isVisible(this)) {
           this.focus()
         }
       })
     })
 
+    const $ = window.$ || window.jQuery
     Modal._jQueryInterface.call($(target), config, this)
```
Comment (Johann-S) [inline js/src/modal.js]: Without a change here, our Modal continue to depend on jQuery

**Review by Johann-S — CHANGES_REQUESTED** [review]

**Comment by Johann-S** [comment]
You made a very good work here @alekitto but they are few changes here before I can merge your PR 👍

File: js/src/util.js [hunk 6]
```diff
@@ -64,7 +75,7 @@ const Util = (() => {
     },
 
     supportsTransitionEnd() {
-      return Boolean(transition)
+      return Boolean(EventHandler.getBrowserTransitionEnd())
```
Comment (Johann-S) [inline js/src/util.js]: Not a big fan here because `EventHandler.getBrowserTransitionEnd()` will be called each time `Util.supportsTransitionEnd()` is called

File: js/src/dom/eventHandler.js (hunk 1 above)
Comment (Herst) [reply js/src/dom/eventHandler.js]: `const` does not work in IE10

Comment (XhmikosR) [reply js/src/dom/eventHandler.js]: We use babel.

**Comment by Johann-S** [comment]
I updated the base branch with `v4-dev`, so that's why they are other commits

**Comment by alekitto** [comment]
Ok, i've rebased my branch with upstream. Now working on your comments..

File: js/src/dom/eventHandler.js (hunk 2 above)
Comment (alekitto) [reply js/src/dom/eventHandler.js]: You're right. Fixed.

File: js/src/modal.js (hunk 5 above)
Comment (alekitto) [reply js/src/modal.js]: I've missed it. I've just changed it.

File: js/src/dom/eventHandler.js (hunk 1 above)
Comment (alekitto) [reply js/src/dom/eventHandler.js]: The `off` implementation was not complete and i think i've messed up some things. Now working on a cleaner and more jQuery-like solution.

File: js/src/util.js [hunk 7]
```diff
@@ -64,7 +77,11 @@ const Util = (() => {
     },
 
     supportsTransitionEnd() {
-      return Boolean(transition)
+      if (transition === null) {
```
Comment (Johann-S) [inline js/src/util.js]: Here you add a check if transition is set or not, you should revert this change it was better before that, because they aren't any check done each time we call this method

**Review by Johann-S — CHANGES_REQUESTED** [review]

File: js/tests/unit/modal.js [hunk 8]
```diff
@@ -626,16 +635,19 @@ $(function () {
...
+        assert.notOk(event.defaultPrevented, 'navigating to href will happen')
 
         setTimeout(function () {
-          assert.ok(event.isDefaultPrevented(), 'model shown instead of navigating to href')
+          assert.ok(evt.defaultPrevented, 'model shown instead of navigating to href')
           done()
         }, 1)
       })
-      .trigger('click')
+    $('#test')[0].dispatchEvent(evt)
```
Comment (Johann-S) [inline js/tests/unit/modal.js]: if you want you can use `EventHandler` here

File: js/src/util.js (hunk 7 above)
Comment (alekitto) [reply js/src/util.js]: Cyclic import (util -> eventHandler -> util) was causing me troubles yesterday, but now seems working correctly. Probably it was another error.

**Comment by Johann-S** [comment]
Oups last thing I forgot, can you update our visual modal test ? please

**Review by Johann-S — APPROVED** [review]

**Comment by alekitto** [comment]
Ok, EventHandler off method has been rewritten to support namespaced events (`click.namespace`), native events without namespace (`click` removes `click.namespace`) and namespace only (`.namespace` removes `click.namespace` and `keydown.namespace`).
Visual modal test has been updated

**Comment by Johann-S** [comment]
I don't understand why you made that change on `off` because `off` already support namespaced events 🤔 
What's the difference between what you've done and what we had ? and why you did that change ?

File: js/tests/unit/dom/eventHandler.js [hunk 9]
```diff
@@ -0,0 +1,233 @@
+$(function () {
+  'use strict'
+
+  QUnit.module('event handler')
+
+  QUnit.test('should be defined', function (assert) {
+    assert.expect(1)
+    assert.ok(EventHandler, 'button method is defined')
```
Comment (Johann-S) [inline js/tests/unit/dom/eventHandler.js]: that's not our `button` plugin 😄

**Review by Johann-S — CHANGES_REQUESTED** [review]

**Comment by alekitto** [comment]
The `off` method did not remove events if event type like `click.dismiss.bs.modal` was passed.  This would lead to events not removed in the `hide` method, especially from `document` and `window` nodes. Additionally, non namespaced events was removed only if an handler function was specified as third argument

**Comment by Johann-S** [comment]
Ok thank you for the explanation 👍 and yep the second point wasn't implemented thank you

**Comment by alekitto** [comment]
Yes, now `EventHandler.off(element, 'click.bs.alert')` is working as expected

File: js/tests/unit/dom/eventHandler.js (hunk 9 above)
Comment (alekitto) [reply js/tests/unit/dom/eventHandler.js]: Whoops

**Review by Johann-S — APPROVED** [review]

**Comment by arcanedev-maroc** [comment]
About Modal events, can i suggest a feature request ?

How about adding a public method to listen a modal event:

```js
class Modal {
    //...

    on(event, callback) {
        this._element.addEventListener(`${event}${EVENT_KEY}`, callback);    
    }

    //...
}
```

So instead of:

```js
var myModalEl = document.getElementById('myModal')
myModalEl.addEventListener('hidden.bs.modal', function (e) {
    // do something...
})

// OR

let myModal = new bootstrap.Modal(document.getElementById('myModal'));
myModal._element.addEventListener('hidden.bs.modal', function (e) {
    // do something...
})
```

We can do something like this:

```js
var myModal = new bootstrap.Modal(document.getElementById('myModal'));
myModal.on('hidden', function (e) {
    // do something...
});
```

I don't know if it's going to have a conflict with jquery but we can change the method name `on` into something else like `listen('hidden', ...)`.

I can create a PR for this feature if it's OK :+1:

**Comment by alekitto** [comment]
My 2 cents: This could be an interesting feature, but I think that should be implemented for all the components.
For example:

```js
const myCollapse = new bootstrap.Collapse(document.getElementById('test-collapse'));
myCollapse.on('shown', e => { ... });
```

could be equally good and useful.

@arcanedev-maroc probably you should open a new issue proposing a these changes to the event system of the components to better discuss pros and cons of this new design.

**Comment by arcanedev-maroc** [comment]
@alekitto DONE :+1:
